core/main_loop.py

```diff
@@ -31,6 +31,26 @@
   #from osh.cmd_exec import Executor
 
 
+def _ExecutePromptCommand(ex):
+  # type: (Any) -> None
+  command = None
+  prompt_var = ex.mem.GetVar('PROMPT_COMMAND')
+  if prompt_var.tag == value_e.Undef:
+    return
+  elif prompt_var.tag == value_e.AssocArray:
+    command = prompt_var.d['0']
+  elif prompt_var.tag == value_e.StrArray:
+    command = prompt_var.strs[0]
+  elif prompt_var.tag == value_e.Str:
+    command = prompt_var.s
+
+  # save this so PROMPT_COMMAND can't set $?
+  old_status = ex.LastStatus()
+  arg_vec = arg_vector(['PROMPT_COMMAND', command], [0, 0])
+  ex._Eval(arg_vec)
```

Comment:
Please use PushStatusFrame() / PopStatusFrame(), and use a try/finally so that it gets restored in the case of exceptions (which happen a lot in OSH!)

It would be nice to test that in the spec tests, though I'm not sure if it's possible to tickle.